### DISCUSSION TITLE
app api routes can't return xml content?

### DISCUSSION THREAD

**Discussion by emmgfx** [vercel/next.js#52831] [App Router] [answered]
In pages router, we could return a `text/xml` content doing something like this:

```javascript
export async function getServerSideProps({ res }) {
  // ...
  res.setHeader('Content-Type', 'text/xml');
  res.write(xw.toString());
  res.end();
```

But in app router with the route handlers, i've only found this:

```javascript
import { NextResponse } from 'next/server'
 
export async function GET() {
  // ...
  return NextResponse.json({ data })
}
```

We lost the ability to return a xml file?

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Not at all:

```ts
// src/app/api/sitemap/route.ts
export const dynamic = "force-dynamic";

export async function GET() {
  const sitemapResponse = await fetch("https://www.sitemaps.org/sitemap.xml");

  const xml = await sitemapResponse.text();

  const response = new Response(xml, {
    status: 200,
    statusText: "ok",
  });

  response.headers.append("content-type", "text/xml");

  return response;
}
```

Or I think this works too:

```ts
export async function GET() {
  return fetch("https://www.sitemaps.org/sitemap.xml"); // don't even need to await
}
```

**Reply by emmgfx** [reply]
Thank you @icyJoseph , you're 100% right!

Would you mind let me know where do you found this? I looked for this in the docs, issues, discussions, and didn't found anything :(

**Reply by icyJoseph** [reply]
Right, perhaps they should add some more examples to the docs.

The key thing here is that, the actual power of App Router Route Handlers (say that twice) comes from the Web Fetch API. The handlers receive a Request object, and return a Response object:

- https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API

Most things you can do with those, you can do within a Route Handler, such as responding with text, or JSON, or binary data. 

The sleight of hand is that I've done this before in API Routes (pages router), and for me it was just about removing code, because the Web Fetch API can do so much already:

- https://github.com/vercel/next.js/discussions/52588 Here I helped someone with that

You see my answer was verbose, but then the OP pointed out that it could even be simpler, which is what I answered to you. 

Like now I see the docs and I found this:

```ts
import { cookies } from 'next/headers'
 
export async function GET(request: Request) {
  const cookieStore = cookies()
  const token = cookieStore.get('token')
 
  return new Response('Hello, Next.js!', {
    status: 200,
    headers: { 'Set-Cookie': `token=${token.value}` },
  })
}
```

Which means that line where I do ` response.headers.append("content-type", "text/xml");`, could've also been passed to the Response constructor.

However the simpler `return await fetch` tops it all, for what is worth, not even the `await` is needed IIRC. All of that because fetch resolves to a Reponse.

**Reply by emmgfx** [reply]
Thank you. Very helpful!
